### ISSUE TITLE
Sometimes response body cut in half.

### ISSUE THREAD

**Issue by NikosEfthias** [sagebind/isahc#154]
In my app sometimes huge json bodies I am trying to read is received half. Is there any limit or configuration I should make to prevent this?

**Comment by sagebind** [comment]
I don't think so, I wouldn't expect that to happen. Maybe a bug in Isahc somewhere? 

If you enable trace logging, when this occurs do any logs stand out as suspicious? Maybe there's something in those logs that can point us in the right direction. 

_Tip: to enable trace logging, hook up a logger crate like env_logger with the `RUST_LOG=isahc=trace` environment variable set._

**Comment by NikosEfthias** [comment]
![image](https://user-images.githubusercontent.com/9097437/74392407-7231ae80-4e20-11ea-8e21-dd3670cea178.png)
I recently had time to review the errors.I pasted the partial log output above. 
For some reason sockets are being closed half which started to happen a lot recently

**Comment by NikosEfthias** [comment]
any ideas why this might be happening @sagebind

**Comment by sagebind** [comment]
Not sure, I've never seen this before. These lines seem curious:

```
TLSv1.3 (OUT), TLS alert, close notify (256):
transfer closed with 12025 bytes remaining to read
```

Not sure what the TLS alert means, does that mean that this is somehow TLS related?

> huge json bodies

How huge is that in bytes? Maybe I can try to perform something similar in a debugger and try to reproduce.

Also, it would help if I knew what version of Isahc you are using.

_Note to self: #77 may have helped here to isolate the logs for a particular failed transfer._

**Comment by NikosEfthias** [comment]
im using version  `0.8.2` I am not exactly sure about the body size I will update here if I get a chance to update the running binary on production.
![image](https://user-images.githubusercontent.com/9097437/74603038-826bb700-50c8-11ea-9afd-09ac4b848703.png)
this is how I produce the request 
and this is how I send it 
![image](https://user-images.githubusercontent.com/9097437/74603067-d8405f00-50c8-11ea-800a-daa3fb057448.png)
![image](https://user-images.githubusercontent.com/9097437/74603095-44bb5e00-50c9-11ea-9815-fba620099c71.png)

**Comment by sagebind** [comment]
Good news, I was finally able to reproduce this! I'm not sure if this is exactly what you're running into, but the log output sure looks similarto what you are seeing. I reproduced it by having `.text()` fail mid-stream with invalid UTF-8 and force the connection to abort.

I took [this UTF-8 decoder stress test file](https://www.cl.cam.ac.uk/~mgk25/ucs/examples/UTF-8-test.txt), duplicated it 100,000 times to make a big version of it (so that downloading wouldn't asynchronously complete while decoding), and then ran this code:

```rust
use isahc::{
    http,
    prelude::*,
};

fn main() {
    env_logger::init();

    http::Request::get("http://localhost:8000/big-bad-utf8.txt")
        .body(())
        .unwrap()
        .send()
        .unwrap()
        .text()
        .unwrap();
}
```

Output:

```
2020-02-26T05:13:58Z TRACE isahc::handler] received 16384 bytes of data
[2020-02-26T05:14:00Z DEBUG isahc::handler] transfer closed with 221452056 bytes remaining to read
[2020-02-26T05:14:00Z DEBUG isahc::handler] Closing connection 0
[2020-02-26T05:14:00Z DEBUG isahc::handler] curl error: [18] Transferred a partial file
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "stream did not contain valid UTF-8" })', src/libcore/result.rs:1188:5
```

What's happening here is that `text()` is streaming the response bytes and decoding it as UTF-8 incrementally. If there are malformed UTF-8 bytes in there, then it will return an error early, potentially causing the transfer to abort and result in the types of logs you see above.

In your code, what sort of error result are you getting back? Are you sure you're not getting an error from `.text()`? Perhaps sometimes the response body contains a malformed character and that is giving you trouble.

I tried lots of other things that didn't help, but I at least I know what this _isn't_. The file size isn't a problem here; Isahc is able to transfer very large ISO files perfectly (and quickly) and the digests of the output are correct, so the problem isn't the size of the payload.

**Comment by sagebind** [comment]
Interestingly, as I test this, I am noticing that enabling or disabling `trace` logs actually effects the results a bit as to how early things abort, which is rather strange. I'm not sure it is related to your issue, but I will look into that to be sure.

**Comment by sagebind** [comment]
Finally found a fix for this in #246! This was a bug caused by uncaught errors during response body (such as timeouts) instead causing an unexplained EOF. A fix for this will be available in the next release.

**Comment by sagebind** [comment]
This is now fixed in the [0.9.11 release](https://github.com/sagebind/isahc/releases/tag/0.9.11)!
